backend/server.js

```diff
@@ -1,29 +1,179 @@
+  }
+});
+
+// Middleware to hash password before new user is saved
+userSchema.pre('save', async function (next) {
+  const user = this;
+
+  if (!user.isModified('password')) {
+    return next();
+  }
+  const salt = bcrypt.genSaltSync();
+  user.password = bcrypt.hashSync(user.password, salt);
+  next();
+});
+
+// Middleware to authenticate user
+const authenticateUser = async (req, res, next) => {
+  try {
+    const accessToken = req.header('Authorization');
+    const user = await User.findOne({ accessToken });
+    if (!user) {
+      throw 'User not found';
+    }
+    req.user = user;
+    next();
+  } catch (err) {
+    const errorMessage = 'Please try logging in again';
+    res.status(401).json({ error: errorMessage });
+  }
+};
+
+// Model
+const User = mongoose.model('User', userSchema);
+
+// PORT=9000 npm start
+const port = process.env.PORT || 8080;
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
 
-// Start defining your routes here
+// Simple usage
+// This is how we have used CORS in previous project, which is a Simple Usage (Enable All CORS Requests).
+// As default the CORS allow all domain to access the server.
+// Access-Control-Allow-Origin: * is shown in the network tab.  The * = all domain has access.
+
+app.use(cors());
+app.use(bodyParser.json());
+
+// GET - list of all endpoints
+const endPointList = require('express-list-endpoints');
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  if (!res) {
+    res
+      .status(404)
+      .send({ error: 'Oops! Something goes wrong. Try again later!' });
+  }
+  res.send(endPointList(app));
+});
+
+// Configuring CORS
+// To limit access to a specific domain, below is an example how to write that in code.
+// We only allow http://localhost:3000 as an origin.
+// app.use(cors());
+// app.use(bodyParser.json());
+
+// var corsOptions = {
+//   origin: 'http://localhost:3000',
+//   optionsSuccessStatus: 200 // some legacy browsers (IE11, various SmartTVs) choke on 204
+// };
+
+// app.get('/', cors(corsOptions), (req, res) => {
+//   res.json('Hello hello');
+// });
+
+// Const's for error messages instead of text in error handling
+const SERVICE_UNAVAILABLE = 'Service unavailable.';
+
+// Error message in case server is down
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next();
+  } else {
+    res.status(503).send({ error: SERVICE_UNAVAILABLE });
+  }
+});
+
+// POST - Signup a user
+app.post('/signup', async (req, res) => {
+  try {
+    const { name, email, password } = req.body;
+    const user = await new User({
+      name,
+      email,
+      password
+    }).save();
+    res.status(201).json({ userId: user._id, accessToken: user.accessToken });
+  } catch (err) {
+    res
+      .status(400)
+      .json({ message: 'Could not create user', errors: err.errors });
+  }
+});
+
+// Login user
+app.post('/login', async (req, res) => {
```

Comment:
/signup and /login are not plurals - when creating a RESTful API we need to think about this. 
For example on line 140 if multiple "login" will be active then it should be written as "logins" or "sessions".

Reply:
Updated to 
/users --> for signup a user
/sessions --> for login a user